presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/ProjectionPushThroughJoin.java

```diff
@@ -0,0 +1,246 @@
+                    assignmentBuilder.put(key, expression);
+                }
+
+                if (!leftAssignments.isEmpty() || !rightAssignments.isEmpty()) {
+                    if (pushdownAll) {
+                        return context.rewrite(node.getSource(), new PushdownContext(leftAssignments, rightAssignments));
+                    }
+                    return new ProjectNode(idAllocator.getNextId(), context.rewrite(node.getSource(), new PushdownContext(leftAssignments, rightAssignments)), assignmentBuilder.build());
+                }
+                return context.defaultRewrite(node);
+            }
+            return buildNode(node, node.getSource().getOutputSymbols(), context);
+        }
+
+        @Override
+        public PlanNode visitJoin(JoinNode node, RewriteContext<PushdownContext> context)
+        {
+            PushdownContext assignment = context.get();
+            if (assignment == null) {
+                return context.defaultRewrite(node);
+            }
+
+            PlanNode left = node.getLeft();
+            PlanNode right = node.getRight();
+            Map<Symbol, Expression> leftAssignments = assignment.getLeftAssignment();
+            Map<Symbol, Expression> rightAssignments = assignment.getRightAssignment();
+            ImmutableMap.Builder<Symbol, Expression> builder = ImmutableMap.builder();
+            builder.putAll(leftAssignments);
+            builder.putAll(rightAssignments);
+
+            if (!leftAssignments.isEmpty()) {
+                if (!canPushDown(leftAssignments, node)) {
+                    return new ProjectNode(idAllocator.getNextId(), node, buildContext(builder.build(), node.getOutputSymbols()));
+                }
+                List<Symbol> leftSymbols = node.getCriteria().stream().map(EquiJoinClause::getLeft).collect(toList());
+                left = context.rewrite(node.getLeft(), new PushdownContext(buildContext(leftAssignments, leftSymbols), new HashMap<>()));
+            }
+
+            if (!rightAssignments.isEmpty()) {
+                if (!canPushDown(rightAssignments, node)) {
+                    return new ProjectNode(idAllocator.getNextId(), node, buildContext(builder.build(), node.getOutputSymbols()));
+                }
+                List<Symbol> rightSymbols = node.getCriteria().stream().map(EquiJoinClause::getRight).collect(toList());
+                right = context.rewrite(node.getRight(), new PushdownContext(new HashMap<>(), buildContext(rightAssignments, rightSymbols)));
+            }
+            return new JoinNode(idAllocator.getNextId(), node.getType(), left, right, node.getCriteria(), node.getFilter(), node.getLeftHashSymbol(), node.getRightHashSymbol());
+        }
+
+        private boolean isPushDownExpression(Expression expression)
+        {
+            return expression instanceof ArithmeticBinaryExpression || expression instanceof ArithmeticUnaryExpression || expression instanceof ComparisonExpression;
```

Comment:
why this preselection?
It seems that function calls can be expensive too so it might be worth pushing it down.